corehq/apps/domain/tests/test_auth_type.py

```diff
@@ -174,3 +180,52 @@ def test_user_agent_beats_header(self):
         self.assertEqual('basic',
                          determine_authtype_from_request(self._mock_request(user_agent='Android',
                                                                             auth_header='Digest whatever')))
+
+
+class TestDetermineAuthTypeFromRequest(SimpleTestCase):
+    """
+    Similar approach to the above test case, but here we use python requests to
+    set the headers
+    """
+
+    def get_django_request(self, auth=None, headers=None):
+        def to_django_header(header_key):
+            # python simple_server.WSGIRequestHandler does basically this:
+            return u'HTTP_' + header_key.upper().replace('-', '_')
+
+        req = (requests.Request(
+            'GET',
+            'https://example.com',
+            auth=auth,
+            headers=headers,
+        ).prepare())
+
+        return RequestFactory().generic(
+            method=req.method,
+            path=req.path_url,
+            data=req.body,
+            **{to_django_header(k): v for k, v in req.headers.items()}
+        )
+
+    def test_basic_auth(self):
+        request = self.get_django_request(auth=requests.auth.HTTPBasicAuth('foo', 'bar'))
+        self.assertEqual('basic', determine_authtype_from_request(request))
+
+    def test_digest_auth(self):
+        request = self.get_django_request(auth=requests.auth.HTTPDigestAuth('foo', 'bar'))
+        # FIXME This is wrong - this should return digest auth
+        self.assertEqual('basic', determine_authtype_from_request(request))
```

Comment:
@snopoke Any thoughts on this?  Yesterday @proteusvacuum and I figured out that anything using `login_or_digest_or_basic_or_apikey` or `login_or_digest_or_basic_or_apikey_or_token` (with the current default parameter of `BASIC`) doesn't actually work with digest auth.  Requests from a J2ME device and tastypie endpoints are special-cased to default to digest auth, so they do work fine, as far as we know.

The reason this doesn't work is that to use digest auth, the client first sends an unauthenticated request, and the server responds with a 401 status code and some parameters needed to construct a digest auth request.  Thus we have no way of knowing which auth method the client intends to use - we can't distinguish between an initial digest auth request and a normal unauthenticated request.

Defaulting to digest should work, as a basic auth request does include parameters on the first request.  Still, I'm not confident that there aren't odd use-cases that would break with a change like that.

Reply:
@emord FYI in case this falls under your security umbrella

Reply:
Interesting, that all makes sense but then how does this work for Android when it makes requests to endpoints like `search`, `claim`, `fetch_key_records` etc?

Reply:
`determine_authtype_from_request` explicitly returns `BASIC` when `"Android"` is in the user agent string, which I'm guessing means Android is always expected to use basic auth.